src/components/views/elements/Validation.js

```diff
@@ -0,0 +1,102 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import classNames from 'classnames';
+
+/**
+ * Creates a validation function from a set of rules describing what to validate.
+ *
+ * @param {String} description
+ *     Summary of the kind of value that will meet the validation rules. Shown at
+ *     the top of the validation feedback.
+ * @param {Object} rules
+ *     An array of rules describing how to check to input value. Each rule in an object
+ *     and may have the following properties:
+ *     - `key`: A unique ID for the rule. Required.
+ *     - `regex`: A regex used to determine the rule's current validity. Required.
+ *     - `valid`: Text to show when the rule is valid. Only shown if set.
+ *     - `invalid`: Text to show when the rule is invalid. Only shown if set.
+ * @returns {Function}
+ *     A validation function that takes in the current input value and returns
+ *     the overall validity and a feedback UI that can be rendered for more detail.
+ */
+export default function withValidation({ description, rules }) {
+    return function onValidate(value) {
+        // TODO: Hide on blur
+        // TODO: Re-run only after ~200ms of inactivity
+        if (!value) {
+            return {
+                valid: null,
+                feedback: null,
+            };
+        }
+
+        const results = [];
+        let valid = true;
+        if (rules && rules.length) {
+            for (const rule of rules) {
+                if (!rule.key || !rule.regex) {
+                    continue;
+                }
+                const ruleValid = rule.regex.test(value);
+                valid = valid && ruleValid;
+                if (ruleValid && rule.valid) {
+                    // If the rule's result is valid and has text to show for
+                    // the valid state, show it.
+                    results.push({
+                        key: rule.key,
+                        valid: true,
+                        text: rule.valid,
+                    });
+                } else if (!ruleValid && rule.invalid) {
+                    // If the rule's result is invalid and has text to show for
+                    // the invalid state, show it.
+                    results.push({
+                        key: rule.key,
+                        valid: false,
+                        text: rule.invalid,
+                    });
+                }
+            }
+        }
+
+        let details;
+        if (results && results.length) {
+            details = <ul className="mx_Validation_details">
+                {results.map(result => {
+                    const classes = classNames({
+                        "mx_Validation_detail": true,
+                        "mx_Validation_valid": result.valid,
+                        "mx_Validation_invalid": !result.valid,
+                    });
+                    return <li key={result.key} className={classes}>
```

Comment:
nit: a bit weird that markup/components will get stored in the the state.feedback of the field. I'd rather expect a description object that Field or a subcomponent renders. Not new to this PR though.

Reply:
Right, we are doing a bit of an unusual thing with design of `Field` and the validation API. I've been quite unsure about it so far, and prototyped quite a few other approaches, such as having  a separate component that contains the `Field`, but none of them have really felt great, so I've just been going with this to get something in place.